File: packages/wallet-native-sdk/android/CoinbaseWalletNativeSDK/src/main/java/com/coinbase/android/nativesdk/task/TaskManager.kt
package com.coinbase.android.nativesdk.task

import com.coinbase.android.nativesdk.CoinbaseWalletSDKError
import com.coinbase.android.nativesdk.message.request.RequestMessage
import com.coinbase.android.nativesdk.message.response.ResponseContent
import com.coinbase.android.nativesdk.message.response.ResponseHandler
import com.coinbase.android.nativesdk.message.response.ResponseMessage
import com.coinbase.android.nativesdk.message.response.ResponseResult
import java.util.Date

internal class TaskManager {
    private val tasks = HashMap<String, Task>()

    fun registerResponseHandler(message: RequestMessage, handler: ResponseHandler) {
        tasks[message.uuid] = Task(
            request = message,
            handler = handler,
            timestamp = message.timestamp
        )
    }

    fun handleResponse(message: ResponseMessage): Boolean {
        val requestId: String
        val result: ResponseResult = when (val response = message.content) {
            is ResponseContent.Response -> {
                requestId = response.requestId
                Result.success(response.values)
            }
            is ResponseContent.Failure -> {
                requestId = response.requestId
                Result.failure(CoinbaseWalletSDKError.WalletReturnedError(response.description))
            }
        }

        val task = tasks[requestId] ?: return false

        task.handler(result)
        tasks.remove(requestId)
        return true
    }

    fun reset(){
        tasks.clear()
    }
}

File: native/ios/Task/TaskManager.swift
//
//  TaskManager.swift
//  WalletSegue
//
//  Created by Jungho Bang on 6/9/22.
//

import Foundation

@available(iOS 13.0, *)
class TaskManager {
    private var tasks = [UUID: Task]()
    
    func registerResponseHandler(
        for requestMessage: RequestMessage,
        _ handler: @escaping ResponseHandler
    ) {
        let uuid = requestMessage.uuid
        tasks[uuid] = Task(
            request: requestMessage,
            handler: handler,
            timestamp: Date()
        )
    }
    
    @discardableResult func handleResponseMessage(_ message: ResponseMessage) -> Bool {
        let requestId: UUID
        let result: ResponseResult
        switch message.content {
        case .response(let response):
            requestId = response.requestId
            result = .success(response)
        case .error(let error):
            requestId = error.requestId
            result = .failure(CoinbaseWalletSDKError.walletReturnedError(error.description))
        }
        
        guard let task = tasks[requestId] else {
            return false
        }
        
        task.handler(result)
        tasks.removeValue(forKey: requestId)
        return true
    }
    
}
